cocos/core/scene-graph/node.ts

```diff
@@ -596,24 +596,25 @@ export class Node extends BaseNode implements CustomSerializable {
         let j = 0;
         let l = 0;
         let cur: this;
-        let children:this[];
-        let hasChangedFlags = 0;
+        let children: this[];
         const childDirtyBit = dirtyBit | TransformBit.POSITION;
 
         dirtyNodes[0] = this;
 
         while (i >= 0) {
             cur = dirtyNodes[i--];
-            hasChangedFlags = cur.hasChangedFlags;
-            if (cur.isValid && (cur._dirtyFlags & hasChangedFlags & dirtyBit) !== dirtyBit) {
-                cur._dirtyFlags |= dirtyBit;
+            if (cur._flagChangeVersion !== globalFlagChangeVersion || (cur._dirtyFlags & cur._hasChangedFlags & dirtyBit) !== dirtyBit) {
+                if (cur.isValid) {
```

Comment:
These two `if` can be merged?

Reply:
These two if can be merged?
>>> Sure. And I thought the cur.isValid checking should be after the flags checking. Because Node.isValid === false only happend when node is destroyed.

Reply:
`.isValid` can be set as the last condition.

Reply:
Please merge the conditions and reduce unneccessary code block.

Reply:
It's merged.